rustls/src/crypto/tls12.rs

```diff
@@ -1,4 +1,65 @@
-pub(crate) fn prf(out: &mut [u8], hmac_key: &dyn super::hmac::Key, label: &[u8], seed: &[u8]) {
+use super::hmac;
+use super::ActiveKeyExchange;
+use crate::error::Error;
+
+use alloc::boxed::Box;
+
+/// An instantiation of the TLS1.2 PRF with a specific, implicit hash function.
+///
+/// See the definition in [RFC5246 section 5](https://www.ietf.org/rfc/rfc5246.txt).
+///
+/// See [`PrfUsingHmac`] as a route to implementing this trait with just
+/// an implementation of [`hmac::Hmac`].
+pub trait Prf: Send + Sync {
+    /// Completes the given key exchange, and then uses the result to compute
+    /// `PRF(secret, label, seed)`, writing the reuslt into `output`.
+    ///
+    /// This can fail only if the key exchange fails.
```

Comment:
> Is this true in general for a PRF construction, or just the HMAC provider?

I think it's true of the construction: it should be a straightforward & deterministic computation, and shouldn't involve memory allocation, IO, etc. Though I'm sure someone has managed to make HMAC or hashing allocate memory -- I think what I'm trying to say is: those implementations would have to panic if they cannot do the computation.

> Relatedly, I'm not sure I understand the reason to prefer taking the Box<dyn ActiveKeyExchange> instead of having the caller complete the KX and pass in the SharedSecret for the PRF to use. 

Background for that choice is https://github.com/rustls/rustls/issues/1540#issuecomment-1768982744

Reply:
>  I think what I'm trying to say is: those implementations would have to panic if they cannot do the computation.

:+1: Ok, I can buy that.

> Background for that choice is https://github.com/rustls/rustls/issues/1540#issuecomment-1768982744

Ahhh, that makes sense. Thanks.